File: internal/stats/verdict.go
package stats

import "time"

// Grade is an overall connection quality rating.
type Grade int

const (
	Excellent Grade = iota
	Good
	Degraded
	Poor
)

func (g Grade) String() string {
	switch g {
	case Excellent:
		return "excellent"
	case Good:
		return "good"
	case Degraded:
		return "degraded"
	default:
		return "poor"
	}
}

// Verdict grades the connection. Loss and jitter dominate the grade because
// they hurt interactive traffic more than raw latency does; latency adjusts
// the grade within a stable connection.
func (t *Tracker) Verdict() (Grade, string) {
	if t.Recv() == 0 {
		return Poor, "no replies received"
	}

	loss := t.Loss()
	jitter := t.Jitter()
	avg := t.Avg()

	switch {
	case loss > 0.05:
		return Poor, "heavy packet loss"
	case loss > 0.01:
		return Degraded, "intermittent packet loss"
	case jitter > 50*time.Millisecond:
		return Degraded, "unstable latency"
	case avg > 300*time.Millisecond:
		return Degraded, "very high latency"
	case loss > 0 || jitter > 20*time.Millisecond || avg > 150*time.Millisecond:
		return Good, "stable with minor variance"
	default:
		return Excellent, "stable, low latency"
	}
}

File: src/verdict.rs
//! Connection quality grading.

use std::fmt;
use std::time::Duration;

use crate::stats::Tracker;

/// An overall connection quality rating.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Grade {
    Excellent,
    Good,
    Degraded,
    Poor,
}

impl fmt::Display for Grade {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(match self {
            Grade::Excellent => "excellent",
            Grade::Good => "good",
            Grade::Degraded => "degraded",
            Grade::Poor => "poor",
        })
    }
}

impl Tracker {
    /// Grades the connection. Loss and jitter dominate the grade because they
    /// hurt interactive traffic more than raw latency does; latency adjusts
    /// the grade within a stable connection.
    pub fn verdict(&self) -> (Grade, &'static str) {
        if self.recv() == 0 {
            return (Grade::Poor, "no replies received");
        }

        let loss = self.loss();
        let jitter = self.jitter();
        let avg = self.avg();

        if loss > 0.05 {
            (Grade::Poor, "heavy packet loss")
        } else if loss > 0.01 {
            (Grade::Degraded, "intermittent packet loss")
        } else if jitter > Duration::from_millis(50) {
            (Grade::Degraded, "unstable latency")
        } else if avg > Duration::from_millis(300) {
            (Grade::Degraded, "very high latency")
        } else if loss > 0.0
            || jitter > Duration::from_millis(20)
            || avg > Duration::from_millis(150)
        {
            (Grade::Good, "stable with minor variance")
        } else {
            (Grade::Excellent, "stable, low latency")
        }
    }
}
